File: 2019/02/python/intcode.py
from typing import List

class Intcode:
  def __init__(self, codes: List[int]) -> None:
    self.codes = list(codes)

  def run(self, noun: int, verb: int) -> int:
    position = 0

    self.codes[1] = noun
    self.codes[2] = verb

    while self.codes[position] != 99:
      opcode = self.codes[position]
      lhs = self.codes[position + 1]
      rhs = self.codes[position + 2]
      res = self.codes[position + 3]

      if opcode == 1:
        self.codes[res] = self.codes[lhs] + self.codes[rhs]
      elif opcode == 2:
        self.codes[res] = self.codes[lhs] * self.codes[rhs]

      position += 4
      
    return self.codes[0]

File: 2019/02/ruby/intcode.rb
class Intcode
  attr_reader :codes

  def initialize(codes)
    @codes = codes.dup
  end

  def run(noun, verb)
    codes[1] = noun
    codes[2] = verb

    position = 0

    while codes[position] != 99
      opcode = codes[position]
      position_lhs = codes[position + 1]
      position_rhs = codes[position + 2]
      position_result = codes[position + 3]

      case opcode
      when 1
        codes[position_result] = codes[position_lhs] + codes[position_rhs]
      when 2
        codes[position_result] = codes[position_lhs] * codes[position_rhs]
      end

      position += 4
    end

    codes[0]
  end
end
